src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs

```diff
@@ -207,5 +240,41 @@ public Task UnloadAsync()
 
             return Task.CompletedTask;
         }
+
+        // true means the project system plans to call NominateProjectAsync in the future.
+        public bool HasPendingNomination()
+        {
+            return _pendingNomination;
+        }
+
+        // NuGet calls this method to wait project to nominate restoring
+        // If the project has no pending restore data, it will return a completed task.
+        // Otherwise a task which will be completed once the project norminate the next restore
+        // the task will be cancelled, if the project system decide it no longer need restore (for example: the restore state has no change)
+        // the task will be failed, if the project system runs into a problem, so it cannot get correct data to norminate a restore (DT build failed)
+        public Task WhenRestoreNominated()
+        {
+            if (_pendingNomination == false)
+            {
+                return Task.CompletedTask;
+            }
+            else
+            {
+                // task which will be completed once the project norminate the next restore (_norminateNextRestore == 1)
+                JoinableTask joinableTask = JoinableFactory.RunAsync(() =>
+                {
+                    while (true)
```

Comment:
this is not correct. you should never write a code to loop checking a state, and return when it matches some condition. 

I guess it is correct to call HasPendingNomination(), and if it is false, return a completed task. But the rest is more complex. 

For now, to prevent a race condition, I think we should create a new class member using AsyncAutoResetEvent (no inline awaiters), and just return WaitAsync() here as its return value. Certainly, we won't be able to set a fault to this returned task, but since NuGet doesn't use it, we can resolve it later. (which means we need repeat some work in AsyncAutoResetEvent, which I think we should not do for now. Just focus to get this right.

Now inside RestoreAsync, after it calls RestoreCoreAsync, it should record restored version first. After that, it should set the AsyncAutoResetEvent.

Reply:
Oh, i realized that my statement is not accurate here. Both restore version & AsyncAutoResetEvent need be triggered after calling RestoreCoreAsync (or whatever starts the restore), but it should not wait it to finish. You have to call and retain the return task, and wait for it only after we set the events.  Otherwise, we will wait NuGet restore to finish & NuGet will wait us to say normination is done, which will lead a deadlock. Our flag is only to say we have passed the data to NuGet to restore, but not the NuGet restore is done.

Reply:
> we won't be able to set a fault to this returned task

Could `TaskCompletionSource` be used instead of `AsyncAutoResetEvent` so that fault/cancellation is propagated?

Reply:
I thought AsyncAutoResetEvent might make it a little bit easier to get the code right. As long as we know the TaskCompletionSource need be recreated on each waiting, it might be a better choice in the end.

There are a few critical things to get this thing right:

1, the project might be already changed when we process the dataflow and call RestoreCoreAsync, we need decide whether we should verify & skip it.  I think we should, because otherwise, HasPendingNomination will be true even when we call RestoreCoreAsync, and it will be odd to not complete the waiting task.

2, it is a race condition that project can get dirty right after RestoreCoreAsync is called, and when we complete the task. So the key point:
 a, WhenNorminating is called, if we need wait, we need create TaskCompletionSource (if it is not yet created), and also record the version of project we are waiting for (and need update even we reuse TaskCompletionSource from an earlier waiting task).  The version/CompletionSource need be protected within a same critical section. Right after RestoreCoreAsync is called (not waiting it to complete), it need go through the lock, and check the version with the version earlier waiting task is waiting for. If it is beyond the original value, it completes the task, otherwise, it leaves the task there.  That will prevent the race condition described here.